[Filebeat][New Fileset] Cisco AMP

P1llus opened the pull request:
## What does this PR do?

This PR is a WIP for adding Cisco AMP fileset to the cisco module.

Cisco AMP allows collection of endpoint events from their Events API documented  [here](https://api-docs.amp.cisco.com/api_actions/details?api_action=GET+%2Fv1%2Fevents&api_host=api.amp.cisco.com&api_resource=Event&api_version=v1)

## Why is it important?

Adds support for another very commonly used Cisco product.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Requires #22320

P1llus commented:
The initial PR is created as a draft for easy sharing, its current state is that it is waiting for #22320 and is still missing core components like documentation and a finished ingest pipeline. Should finalize this right after the above PR has been merged to master.

elasticmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22768/detail/PR-22768/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22768/detail/PR-22768/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22768/detail/PR-22768/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22768/detail/PR-22768/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_22768.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `Pull request #22768 opened`

* Start Time: 2020-11-26T05:02:59.121+0000

* Duration: 51 min 14 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 1978  |
| Skipped      | 259 |
| Total        | 2238   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22768/detail/PR-22768/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
#### `Build&Test / x-pack/filebeat-build / test_fileset_file_014_cisco – x-pack.filebeat.tests.system.test_xpack_modules.XPackTest`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Exception: Key 'cisco.amp.date' found in event is not documented! 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 a = (<test_xpack_modules.XPackTest testMethod=test_fileset_file_014_cisco>,)

    @wraps(func)
    def standalone_func(*a):
>       return func(*(a + p.args), **p.kwargs)

../../build/ve/docker/lib/python3.7/site-packages/parameterized/parameterized.py:518: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../filebeat/tests/system/test_modules.py:99: in test_fileset_file
    cfgfile=cfgfile)
../../filebeat/tests/system/test_modules.py:183: in run_on_file
    self.assert_fields_are_documented(obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test_xpack_modules.XPackTest testMethod=test_fileset_file_014_cisco>
evt = {'@timestamp': '2020-11-11T09:51:38.000Z', 'agent': {'ephemeral_id': 'b9c75092-706f-41a0-9659-f1c692d215c6', 'id': 'ad...:51:38+00:00', 'detection': 'W32.Trojan.22gp.1201', ...}}, 'destination': {'address': '8.8.8.8', 'ip': '8.8.8.8'}, ...}

    def assert_fields_are_documented(self, evt):
        """
        Assert that all keys present in evt are documented in fields.yml.
        This reads from the global fields.yml, means `make collect` has to be run before the check.
        """
        expected_fields, dict_fields, aliases = self.load_fields()
        flat = self.flatten_object(evt, dict_fields)
    
        def field_pattern_match(pattern, key):
            pattern_fields = pattern.split(".")
            key_fields = key.split(".")
            if len(pattern_fields) != len(key_fields):
                return False
            for i in range(len(pattern_fields)):
                if pattern_fields[i] == "*":
                    continue
                if pattern_fields[i] != key_fields[i]:
                    return False
            return True
    
        def is_documented(key, docs):
            if key in docs:
                return True
            for pattern in (f for f in docs if "*" in f):
                if field_pattern_match(pattern, key):
                    return True
            return False
    
        for key in flat.keys():
            metaKey = key.startswith('@metadata.')
            # Range keys as used in 'date_range' etc will not have docs of course
            isRangeKey = key.split('.')[-1] in ['gte', 'gt', 'lte', 'lt']
            if not(is_documented(key, expected_fields) or metaKey or isRangeKey):
>               raise Exception("Key '{}' found in event is not documented!".format(key))
E               Exception: Key 'cisco.amp.date' found in event is not documented!

../../libbeat/tests/system/beat/beat.py:729: Exception 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY-->


### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22768/detail/PR-22768/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
#### `x-pack/filebeat-build - mage build test`
<ul>
<li>Took 32 min 42 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-22768/runs/1/steps/994/log/?start=0">here</a></li>
<li>Description: <code>mage build test</code></l1>
</ul>
  
  
  
  
#### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-22768/runs/1/steps/1043/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 1'</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-11-26T05:50:09.098Z] Stashed 4 file(s)
[2020-11-26T05:50:09.433Z] + tar --version
[2020-11-26T05:50:09.788Z] + tar --exclude=test-build-artifacts-x-pack/filebeat-build.tgz -czf test-build-artifacts-x-pack/filebeat-build.tgz .
[2020-11-26T05:50:56.522Z] [INFO] Override default googleStorageUpload with some sleep
[2020-11-26T05:50:56.549Z] Sleeping for 1 min 29 sec
[2020-11-26T05:52:25.573Z] [Google Cloud Storage Plugin] Found 1 files to upload from pattern: test-build-artifacts-x-pack/filebeat-build.tgz
[2020-11-26T05:52:26.002Z] [Google Cloud Storage Plugin] Uploading: test-build-artifacts-x-pack/filebeat-build.tgz
[2020-11-26T05:52:36.750Z] + python .ci/scripts/search_system_tests.py
[2020-11-26T05:52:36.774Z] [INFO] system-tests='build/x-pack/filebeat/build/system-tests'. If no empty then let's create a tarball
[2020-11-26T05:52:37.122Z] + tar --version
[2020-11-26T05:52:37.472Z] + tar --exclude=x-pack-filebeat--system-tests-linux.tgz -czf x-pack-filebeat--system-tests-linux.tgz build/x-pack/filebeat/build/system-tests
[2020-11-26T05:53:04.068Z] [INFO] Override default googleStorageUpload with some sleep
[2020-11-26T05:53:04.090Z] Sleeping for 10 sec
[2020-11-26T05:53:14.114Z] [Google Cloud Storage Plugin] Found 1 files to upload from pattern: x-pack-filebeat--system-tests-linux.tgz
[2020-11-26T05:53:14.181Z] [Google Cloud Storage Plugin] Uploading: x-pack-filebeat--system-tests-linux.tgz
[2020-11-26T05:53:20.421Z] Client: Docker Engine - Community
[2020-11-26T05:53:20.421Z]  Version:           19.03.13
[2020-11-26T05:53:20.421Z]  API version:       1.40
[2020-11-26T05:53:20.421Z]  Go version:        go1.13.15
[2020-11-26T05:53:20.421Z]  Git commit:        4484c46d9d
[2020-11-26T05:53:20.421Z]  Built:             Wed Sep 16 17:02:36 2020
[2020-11-26T05:53:20.421Z]  OS/Arch:           linux/amd64
[2020-11-26T05:53:20.421Z]  Experimental:      false
[2020-11-26T05:53:20.421Z] 
[2020-11-26T05:53:20.421Z] Server: Docker Engine - Community
[2020-11-26T05:53:20.421Z]  Engine:
[2020-11-26T05:53:20.421Z]   Version:          19.03.13
[2020-11-26T05:53:20.421Z]   API version:      1.40 (minimum version 1.12)
[2020-11-26T05:53:20.421Z]   Go version:       go1.13.15
[2020-11-26T05:53:20.421Z]   Git commit:       4484c46d9d
[2020-11-26T05:53:20.421Z]   Built:            Wed Sep 16 17:01:06 2020
[2020-11-26T05:53:20.421Z]   OS/Arch:          linux/amd64
[2020-11-26T05:53:20.421Z]   Experimental:     false
[2020-11-26T05:53:20.421Z]  containerd:
[2020-11-26T05:53:20.421Z]   Version:          1.3.7
[2020-11-26T05:53:20.421Z]   GitCommit:        8fba4e9a7d01810a393d5d25a3621dc101981175
[2020-11-26T05:53:20.421Z]  runc:
[2020-11-26T05:53:20.421Z]   Version:          1.0.0-rc10
[2020-11-26T05:53:20.421Z]   GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
[2020-11-26T05:53:20.421Z]  docker-init:
[2020-11-26T05:53:20.421Z]   Version:          0.18.0
[2020-11-26T05:53:20.421Z]   GitCommit:        fec3683
[2020-11-26T05:53:26.318Z] Failed in branch x-pack/filebeat-build
[2020-11-26T05:53:26.582Z] Stage "Packaging" skipped due to earlier failure(s)
[2020-11-26T05:53:31.234Z] [INFO] unstashV2: JOB_GCS_BUCKET is set. bucket param got precedency instead.
[2020-11-26T05:53:31.263Z] [INFO] unstashV2: JOB_GCS_CREDENTIALS is set. credentialsId param got precedency instead.
[2020-11-26T05:53:31.335Z] [Google Cloud Storage Plugin] Found 1 files to download from pattern: gs://beats-ci-temp/Beats/beats/PR-22768-1/source/source.tgz
[2020-11-26T05:53:31.355Z] [Google Cloud Storage Plugin] Downloading: Beats/beats/PR-22768-1/source/source.tgz to local path: /var/lib/jenkins/workspace/Beats_beats_PR-22768/source.tgz
[2020-11-26T05:53:41.289Z] + tar --version
[2020-11-26T05:53:41.607Z] + tar -xpf source.tgz
[2020-11-26T05:53:51.936Z] + rm source.tgz
[2020-11-26T05:53:52.030Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-22768/src/github.com/elastic/beats
[2020-11-26T05:53:52.060Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-22768/src/github.com/elastic/beats/x-pack-filebeat-Lint-1606367893836
[2020-11-26T05:53:52.158Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-22768/src/github.com/elastic/beats/x-pack-filebeat-windows-windows-2019-1606368303233
[2020-11-26T05:53:52.250Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-22768/src/github.com/elastic/beats/x-pack-filebeat-build-1606369808839
[2020-11-26T05:53:52.656Z] + cat
[2020-11-26T05:53:52.656Z] + /usr/local/bin/runbld ./runbld-test-reports --job-name elastic+beats+pull-request
[2020-11-26T05:53:52.656Z] Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
[2020-11-26T05:53:57.937Z] runbld>>> runbld started
[2020-11-26T05:53:57.937Z] runbld>>> 1.6.12/f45d832f2ba0aa2722ab4ec1fda8ad140f027f8b
[2020-11-26T05:53:59.321Z] runbld>>> The following profiles matched the job 'elastic+beats+pull-request' in order of occurrence in the config (last value wins).
[2020-11-26T05:53:59.321Z] runbld>>> Matches in the system config:
[2020-11-26T05:53:59.321Z] runbld>>> - Matched ^elastic\+beats
[2020-11-26T05:53:59.321Z] runbld>>> - Matched ^elastic\+beats\+pull-request
[2020-11-26T05:54:00.702Z] runbld>>> Debug logging enabled.
[2020-11-26T05:54:00.702Z] runbld>>> Storing result
[2020-11-26T05:54:00.702Z] runbld>>> Store result: created {:total 2, :successful 2, :failed 0} 1
[2020-11-26T05:54:00.702Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1597739501209/t/20201126055400-3EDFB741
[2020-11-26T05:54:00.702Z] runbld>>> Adding system facts.
[2020-11-26T05:54:01.642Z] runbld>>> Adding vcs info for the latest commit:  b521b9bc560df153415f7ed3c58bf4adf57dc30a
[2020-11-26T05:54:01.642Z] runbld>>> >>>>>>>>>>>> SCRIPT EXECUTION BEGIN >>>>>>>>>>>>
[2020-11-26T05:54:01.642Z] runbld>>> Adding /usr/lib/jvm/java-8-openjdk-amd64/bin to the path.
[2020-11-26T05:54:01.642Z] + echo 'Processing JUnit reports with runbld...'
[2020-11-26T05:54:01.642Z] Processing JUnit reports with runbld...
[2020-11-26T05:54:01.902Z] runbld>>> <<<<<<<<<<<< SCRIPT EXECUTION END <<<<<<<<<<<<
[2020-11-26T05:54:01.902Z] runbld>>> DURATION: 14ms
[2020-11-26T05:54:01.902Z] runbld>>> STDOUT: 40 bytes
[2020-11-26T05:54:01.902Z] runbld>>> STDERR: 49 bytes
[2020-11-26T05:54:01.902Z] runbld>>> WRAPPED PROCESS: SUCCESS (0)
[2020-11-26T05:54:01.902Z] runbld>>> Searching for build metadata in /var/lib/jenkins/workspace/Beats_beats_PR-22768
[2020-11-26T05:54:02.842Z] runbld>>> Storing build metadata: 
[2020-11-26T05:54:02.842Z] runbld>>> Adding test report.
[2020-11-26T05:54:02.842Z] runbld>>> Searching for junit test output files with the pattern: TEST-.*\.xml$ in: /var/lib/jenkins/workspace/Beats_beats_PR-22768/src/github.com/elastic/beats
[2020-11-26T05:54:03.784Z] runbld>>> Found 6 test output files
[2020-11-26T05:54:04.045Z] runbld>>> Test output logs contained: Errors: 0 Failures: 1 Tests: 2238 Skipped: 244
[2020-11-26T05:54:04.305Z] runbld>>> Storing result
[2020-11-26T05:54:04.305Z] runbld>>> FAILURES: 1
[2020-11-26T05:54:04.565Z] runbld>>> Store result: updated {:total 2, :successful 2, :failed 0} 2
[2020-11-26T05:54:04.565Z] runbld>>> BUILD: https://c150076387b5421f9154dfbf536e5c60.us-west1.gcp.cloud.es.io:9243/build-1597739501209/t/20201126055400-3EDFB741
[2020-11-26T05:54:04.825Z] runbld>>> Email notification disabled by environment variable.
[2020-11-26T05:54:04.825Z] runbld>>> Slack notification disabled by environment variable.
[2020-11-26T05:54:10.178Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-22768/src/github.com/elastic/beats
[2020-11-26T05:54:11.657Z] Running on worker-1244230 in /var/lib/jenkins/workspace/Beats_beats_PR-22768
[2020-11-26T05:54:11.795Z] [INFO] getVaultSecret: Getting secrets
[2020-11-26T05:54:11.864Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-11-26T05:54:14.098Z] + chmod 755 generate-build-data.sh
[2020-11-26T05:54:14.098Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22768/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22768/runs/1 FAILURE 3073582
[2020-11-26T05:54:14.098Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22768/runs/1/steps/?limit=10000 -o steps-info.json
[2020-11-26T05:54:16.354Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22768/runs/1/tests/?status=FAILED -o tests-errors.json
[2020-11-26T05:54:17.051Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-22768/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



## :bug: Flaky test report
:grey_exclamation: There are test failures but not known flaky tests.


<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

### Test stats :test_tube:
| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 1978  |
| Skipped      | 259 |
| Total        | 2238   |


### Genuine test errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22768/detail/PR-22768/1//tests)
:broken_heart: There are test failures but not known flaky tests, most likely a genuine test failure.

* **Name**: `Build&Test / x-pack/filebeat-build / test_fileset_file_014_cisco – x-pack.filebeat.tests.system.test_xpack_modules.XPackTest`


</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

P1llus commented:
Cisco AMP returns the newest event first instead of last, this would require the possibility to get timestamp of first instead of last event added through the httpjson cursor functionality.

P1llus commented:
First MVP is finalized and is now ready for review

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

andrewstucki commented on x-pack/filebeat/module/cisco/amp/test/cisco_amp2.ndjson.log-expected.json:
```diff
@@ -0,0 +1,5676 @@
+        "@timestamp": "2021-01-15T11:59:52.000Z",
+        "cisco.amp.computer.active": true,
+        "cisco.amp.computer.connector_guid": "test_connector_guid",
+        "cisco.amp.computer.external_ip": "8.8.8.8",
```
This should also be under `related.ip`

andrewstucki commented on x-pack/filebeat/module/cisco/amp/test/cisco_amp2.ndjson.log-expected.json:
`related.hosts`

andrewstucki commented on x-pack/filebeat/module/cisco/amp/test/cisco_amp2.ndjson.log-expected.json:
```diff
@@ -0,0 +1,5676 @@
+        "cisco.amp.related.mac": [
+            "87:c2:d9:a2:8c:74"
+        ],
+        "cisco.amp.tactics": [
```
are any of the `threat` fields usable? Looks like they might not be unless ECS moves towards making the whole field set flattened/nested?

P1llus replied:
Yeah unfortunately leaving them there was a conscious choice, the current format only fits a flattened datatype. I will bring it up with the ECS team however, to see what we can do for the future.

andrewstucki commented on x-pack/filebeat/module/cisco/amp/test/cisco_amp2.ndjson.log-expected.json:
```diff
@@ -0,0 +1,5676 @@
+                "mitre_name": "technique",
+                "mitre_url": "https://attack.mitre.org/techniques/T1005",
+                "name": "Data from Local System",
+                "permissions": "",
```
drop this?

P1llus replied:
I believed I had my cleanup painless script I often used, but forgot to add it in, should be fixed now in the upcoming commit.

andrewstucki commented on x-pack/filebeat/module/cisco/amp/test/cisco_amp2.ndjson.log-expected.json:
```diff
@@ -0,0 +1,5676 @@
+                "name": "Scheduled Task/Job",
+                "permissions": "Administrator, SYSTEM, User",
+                "platforms": "Windows, Linux, macOS",
+                "system_requirements": null,
```
drop this?

andrewstucki commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,314 @@
+- set:
+    field: event.severity
+    value: 3
+    if: ctx?.cisco?.amp?.severity == 'Critical'
```
seems like this should be a higher value than "High"?

andrewstucki commented on x-pack/filebeat/module/cisco/amp/test/cisco_amp2.ndjson.log-expected.json:
could we map this to `event.category = "malware"`?

P1llus replied:
Seems like I want to go through the source data again and see if I can add some proper event.category and event.type in here, as its mostly just event.kind at this moment.

andrewstucki commented on x-pack/filebeat/module/cisco/amp/test/cisco_amp.ndjson.log-expected.json:
```diff
@@ -0,0 +1,80 @@
+        "file.name": "HYXiN3hY.exe.part",
+        "file.path": "\\\\?\\C:\\Users\\elastic\\AppData\\Local\\Temp\\HYXiN3hY.exe.part",
+        "fileset.name": "amp",
+        "host.hostname": "testhost",
```
is there any way of filling in the os family? Say by looking at the file path to determine this is windows?

P1llus replied:
I am able to detect windows, but Linux or OSX is going to be more difficult to differentiate.

andrewstucki commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,314 @@
+## ECS Process Mapping ##
+#########################
+- rename:
+    field: cisco.amp.file.parent.process_id
```
so, is this modeling the "parent" of the file? If so I think it should actually map to `process.pid` -- `process.parent` is used for modeling the process ancestry of the process we're currently looking at v. the process that is creating a file (which would be the current process)

P1llus replied:
I believe this should be the actual parent, as this data is used for parent/child relationships between files and processes in the Cisco AMP solution as well.

P1llus replied:
This assumption was correct, adding the changes now.

andrewstucki commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,314 @@
+    ignore_missing: true
+- set:
+    field: network.direction
+    value: outbound
```
this should probably be `egress`

andrewstucki commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,314 @@
+    if: "ctx?.cisco?.amp?.network_info?.nfm?.direction == 'Outgoing connection from'"
+- set:
+    field: network.direction
+    value: outbound
```
this should be `ingress`?

andrewstucki commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,314 @@
+    if: ctx?.source?.ip != null
+- append:
+    field: related.ip
+    value: "{{ destination.ip }}"
```
would we want to run this or any of the other `ip` values through `geoip`?

andrewstucki reviewed (commented):
initial pass

andrewstucki commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,314 @@
+#####################
+- rename:
+    field: cisco.amp.network_info.dirty_url
+    target_field: url.full
```
might be good to run this through the URI parts processor: https://www.elastic.co/guide/en/elasticsearch/reference/master/uri-parts-processor.html and, if it gets merged, subsequently through a `registered_domain` https://github.com/elastic/elasticsearch/pull/67611

P1llus replied:
This will not make the module backwards compatible though, is that okay? I could also add a ignore_failure to true, and at least populate url.full or similar manually if url is already empty (running on holder versions?)

P1llus replied:
Will add it in and keep the fallback to at least move the URL to a ECS field, so if we decide to add backwards compatibility then it will still work.

marc-gr commented on filebeat/docs/modules/cisco.asciidoc:
```diff
@@ -14,6 +14,7 @@ This is a module for Cisco network device's logs and Cisco Umbrella. It includes
 filesets for receiving logs over syslog or read from a file:
 
 - `asa` fileset: supports Cisco ASA firewall logs.
+- `amp` fileset: supports Cisco Umbrella logs.
```
```suggestion
- `amp` fileset: supports Cisco AMP logs.
```

marc-gr commented on filebeat/docs/modules/cisco.asciidoc:
```diff
@@ -444,6 +445,81 @@ Maximum duration before AWS API request will be interrupted. Default to be 120 s
+[float]
+==== `amp` fileset settings
+
+The Cisco AMP fileset focuses on collecting events from your Cisco AMP/Cisco Securi Endpoint API.
```
```suggestion
The Cisco AMP fileset focuses on collecting events from your Cisco AMP/Cisco Secure Endpoint API.
```

marc-gr commented on filebeat/docs/modules/cisco.asciidoc:
```diff
@@ -444,6 +445,81 @@ Maximum duration before AWS API request will be interrupted. Default to be 120 s
+
+The Cisco AMP fileset focuses on collecting events from your Cisco AMP/Cisco Securi Endpoint API.
+
+To configure the Cisco AMP fileset you will need to retrieve your client_id and client_key from the AMP dashboard.
```
```suggestion
To configure the Cisco AMP fileset you will need to retrieve your `client_id` and `client_key` from the AMP dashboard.
```

marc-gr commented on filebeat/docs/modules/cisco.asciidoc:
```diff
@@ -444,6 +445,81 @@ Maximum duration before AWS API request will be interrupted. Default to be 120 s
+
+*`var.input`*::
+
+The input from which messages are read. Supports httpjson(default) and file.
```
Is the `file` input's purpose only testing? If it is the case, maybe there is not much value in documenting it, along with the `var.input` option. WDYT?

marc-gr commented on filebeat/docs/modules/cisco.asciidoc:
```diff
@@ -444,6 +445,81 @@ Maximum duration before AWS API request will be interrupted. Default to be 120 s
+
+The ID for the user account used to access the API.
+
+*`var.api_key`*::
```
Is this making reference to the same field mentioned before as `client_key`?

marc-gr commented on x-pack/filebeat/module/cisco/amp/_meta/fields.yml:
```diff
@@ -0,0 +1,276 @@
+    - name: date
+      type: date
+      description: >
+        The timestamp in ISO8601 format.
```
Will have a specific fixed timezone? If so maybe is good to document it.

P1llus replied:
This field has since been removed, they have 3 different fields for the same timestamp. These are since moved to ECS fields.

marc-gr commented on x-pack/filebeat/module/cisco/amp/_meta/fields.yml:
Are these `*time` fields possible to store as `date` instead?

marc-gr commented on x-pack/filebeat/module/cisco/amp/config/config.yml:
```diff
@@ -0,0 +1,71 @@
+    default: '[[ formatDate (now (parseDuration "-{{ .first_interval }}")) "2006-01-02T15:04:05-07:00" ]]'
+- set:
+    target: url.params.limit
+    value: 100
```
Is this value something that users would like to tweak?

P1llus replied:
I don't see too much value in it, but might be wise to keep it open for users to tweak just in case. It has to exist so we can trigger pagination.

marc-gr commented on x-pack/filebeat/module/cisco/amp/config/config.yml:
```diff
@@ -0,0 +1,71 @@
+  - add_fields:
+      target: ''
+      fields:
+        ecs.version: 1.6.0
```
```suggestion
        ecs.version: 1.7.0
```

marc-gr commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,314 @@
+description: Pipeline for parsing checkpoint firewall logs
```
```suggestion
description: Pipeline for parsing Cisco AMP logs
```

P1llus commented:
The above commit https://github.com/elastic/beats/pull/22768/commits/5a0e1e49be820e4cbc79e970beb43234bc6e25d8 fixes most of the comments made to the PR. Anything not marked resolved are still being worked on.

P1llus commented:
@andrewstucki I added in your feedback to the pipeline, a few small things to note:
1. I can only detect windows based on filepath, its not possible to detect the difference between OSX and Linux, I don't know if you have any ideas or if setting platform to windows is enough to fit most usecases?
2. I added in uri_parts processor, and added a fallback, so that we can continue the discussion around the backwards compatibility in the other issue.

andrewkroh commented on filebeat/docs/modules/cisco.asciidoc:
```diff
@@ -444,6 +445,86 @@ Maximum duration before AWS API request will be interrupted. Default to be 120 s
+To configure the Cisco AMP fileset you will need to retrieve your `client_id` and `api_key` from the AMP dashboard.
+For more information on how to retrieve these credentials, please reference the https://api-docs.amp.cisco.com/api_resources?api_host=api.amp.cisco.com&api_version=v1[Cisco AMP API documentation].
+
+The URL configured for the API depends on which region your AMP is located, currently there is 3 choices:
```
```suggestion
The URL configured for the API depends on which region your AMP is located, currently there are three choices:
```

andrewkroh commented on filebeat/docs/modules/cisco.asciidoc:
```diff
@@ -444,6 +445,86 @@ Maximum duration before AWS API request will be interrupted. Default to be 120 s
+
+*`var.first_interval`*::
+
+How far back you would want to collect events the first time the Filebeat module starts up. Supports amount in hours.
```
Is this really only supported in hours? Normally we accept values like this using `time.Duration` so the value can be given in any format.

P1llus replied:
That is a good point. Initially I set this mostly because we cannot define anything larger than hours, and usually the interval would not need to be less than hour(s). But I can add that it supports minutes and seconds as well.

andrewkroh commented on x-pack/filebeat/module/cisco/amp/config/config.yml:
```diff
@@ -0,0 +1,71 @@
+request.url: {{ .url }}
+request.method: GET
+request.timeout: {{ .request_timeout }}
+request.ssl: {{ .ssl | tojson }}
```
Should this be surrounded with `{{ if .ssl }}` since `ssl` has no default in the manifest? I figured you'd get an error if ssl is not used in the config and you are using the httpjson input.

andrewkroh replied:
Same goes for client_id and api_key.

andrewkroh commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,425 @@
+- append:
+    field: related.ip
+    value: "{{ source.ip }}"
+    if: ctx?.source?.ip != null
```
I think we should use `allow_duplicates: false` on these related.* fields.

andrewkroh commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,425 @@
+    if: ctx?.cisco?.amp?.threat_hunting?.incident_end_time != null
+
+- script:
+    lang: painless
```
Please add a `description` field with a comment about what this does. It's nice to have in the UI.

andrewkroh commented on x-pack/filebeat/module/cisco/amp/ingest/pipeline.yml:
```diff
@@ -0,0 +1,425 @@
+              handleList(x);
+          }
+        }
+      map.values().removeIf(v -> v == null);
```
Is this line's indentation off?
```suggestion
        map.values().removeIf(v -> v == null);
```

adriansr reviewed (approved):
LGTM
